File: libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/DownloadUpdateListenerTest.java
package com.mapbox.services.android.navigation.v5.navigation;

import android.support.annotation.NonNull;

import org.junit.Test;

import java.io.File;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

public class DownloadUpdateListenerTest {

  @Test
  public void onFinishedDownloading_tarIsUnpacked() {
    TileUnpacker tileUnpacker = mock(TileUnpacker.class);
    File file = mock(File.class);
    DownloadUpdateListener downloadUpdateListener = buildDownloadUpdateListener(tileUnpacker);

    downloadUpdateListener.onFinishedDownloading(file);

    verify(tileUnpacker).unpack(any(File.class), any(String.class), any(UnpackProgressUpdateListener.class));
  }

  @Test
  public void onErrorDownloading_offlineErrorIsSent() {
    RouteTileDownloader downloader = mock(RouteTileDownloader.class);
    DownloadUpdateListener downloadUpdateListener = buildDownloadUpdateListener(downloader);

    downloadUpdateListener.onErrorDownloading();

    verify(downloader).onError(any(OfflineError.class));
  }

  @NonNull
  private DownloadUpdateListener buildDownloadUpdateListener(TileUnpacker tileUnpacker) {
    RouteTileDownloader downloader = mock(RouteTileDownloader.class);
    String tilePath = "some/path/";
    String tileVersion = "some-version";
    RouteTileDownloadListener listener = mock(RouteTileDownloadListener.class);
    return new DownloadUpdateListener(
      downloader, tileUnpacker, tilePath, tileVersion, listener
    );
  }

  @NonNull
  private DownloadUpdateListener buildDownloadUpdateListener(RouteTileDownloader downloader) {
    TileUnpacker tileUnpacker = mock(TileUnpacker.class);
    String tilePath = "some/path/";
    String tileVersion = "some-version";
    RouteTileDownloadListener listener = mock(RouteTileDownloadListener.class);
    return new DownloadUpdateListener(
      downloader, tileUnpacker, tilePath, tileVersion, listener
    );
  }
}

File: libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/DownloadUpdateListenerTest.kt
package com.mapbox.services.android.navigation.v5.navigation

import io.mockk.mockk
import io.mockk.verify
import java.io.File
import org.junit.Test

class DownloadUpdateListenerTest {

    @Test
    fun onFinishedDownloading_tarIsUnpacked() {
        val tileUnpacker = mockk<TileUnpacker>(relaxed = true)
        val file = mockk<File>(relaxed = true)
        val downloadUpdateListener = buildDownloadUpdateListener(tileUnpacker)

        downloadUpdateListener.onFinishedDownloading(file)

        verify { tileUnpacker.unpack(any(), any(), any()) }
    }

    @Test
    fun onErrorDownloading_offlineErrorIsSent() {
        val downloader = mockk<RouteTileDownloader>(relaxed = true)
        val downloadUpdateListener = buildDownloadUpdateListener(downloader)

        downloadUpdateListener.onErrorDownloading()

        verify { downloader.onError(any()) }
    }

    private fun buildDownloadUpdateListener(tileUnpacker: TileUnpacker): DownloadUpdateListener {
        val downloader = mockk<RouteTileDownloader>(relaxed = true)
        val tilePath = "some/path/"
        val tileVersion = "some-version"
        val listener = mockk<RouteTileDownloadListener>(relaxed = true)
        return DownloadUpdateListener(
            downloader, tileUnpacker, tilePath, tileVersion, listener
        )
    }

    private fun buildDownloadUpdateListener(downloader: RouteTileDownloader): DownloadUpdateListener {
        val tileUnpacker = mockk<TileUnpacker>(relaxed = true)
        val tilePath = "some/path/"
        val tileVersion = "some-version"
        val listener = mockk<RouteTileDownloadListener>(relaxed = true)
        return DownloadUpdateListener(
            downloader, tileUnpacker, tilePath, tileVersion, listener
        )
    }
}
